src/test/shared/clients/defaultS3Client.test.ts

```diff
@@ -0,0 +1,325 @@
+import { anyFunction, anything, capture, deepEqual, instance, mock, verify, when } from '../../utilities/mockito'
+import * as vscode from 'vscode'
+import {
+    DefaultBucket,
+    DefaultFile,
+    DefaultFolder,
+    DefaultS3Client,
+    FileStreams,
+} from '../../../shared/clients/defaultS3Client'
+import { DEFAULT_DELIMITER, DEFAULT_MAX_KEYS } from '../../../shared/clients/s3Client'
+import { FakeFileStreams } from './fakeFileStreams'
+
+class FakeProgressCaptor {
+    public progress = 0
+
+    public listener(): (loadedBytes: number) => void {
+        return loadedBytes => {
+            this.progress += loadedBytes
+        }
+    }
+}
+
+describe('DefaultS3Client', () => {
+    const partition = 'aws'
+    const region = 'us-west-2'
+    const bucketName = 'bucketName'
+    const outOfRegionBucketName = 'outOfRegionBucketName'
+    const folderPath = 'foo/bar/'
+    const subFolderPath = 'foo/bar/subFolder/'
+    const emptySubFolderPath = 'foo/bar//'
+    const fileKey = 'foo/bar/file.jpg'
+    const fileSizeBytes = 5
+    const fileLastModified = new Date(2020, 5, 4)
+    const fileData = 'fileData'
+    const fileLocation = vscode.Uri.file('/file.jpg')
+    const continuationToken = 'continuationToken'
+    const nextContinuationToken = 'nextContinuationToken'
+
+    let mockS3: S3
+
+    beforeEach(() => {
+        mockS3 = mock()
+    })
+
+    function success<T>(output?: T): Request<T, AWSError> {
+        return {
+            promise: () => Promise.resolve(output),
+            createReadStream() {
+                return FakeFileStreams.readStreamFrom(fileData)
+            },
+        } as Request<any, AWSError>
+    }
+
+    function failure(): Request<any, AWSError> {
+        return {
+            promise: () => Promise.reject(new Error('Expected failure')),
+            createReadStream() {
+                const readStream = FakeFileStreams.readStreamFrom(fileData)
+                readStream.destroy(new Error('Expected failure'))
+                return readStream
+            },
+        } as Request<any, AWSError>
+    }
+
+    function createClient({
+        regionCode = region,
+        partitionId = partition,
+        fileStreams = new FakeFileStreams(),
+    }: { regionCode?: string; partitionId?: string; fileStreams?: FileStreams } = {}): DefaultS3Client {
+        return new DefaultS3Client(regionCode, partitionId, () => Promise.resolve(instance(mockS3)), fileStreams)
+    }
+
+    describe('createBucket', () => {
+        it('creates a bucket', async () => {
+            when(mockS3.createBucket(deepEqual({ Bucket: bucketName }))).thenReturn(success())
+
+            const response = await createClient().createBucket({ bucketName })
+
+            assert.deepStrictEqual(response, {
+                bucket: new DefaultBucket({ partitionId: partition, region, name: bucketName }),
+            })
+        })
+
+        it('throws an S3Error on failure', async () => {
```

Comment:
This doesn't actually check for an S3Error.

Reply:
Yep, good point. We need a better assertion framework so we can do more powerful matching. Looks like it'll be really awkward to match an Error in a type safe way.